README.md

```diff
@@ -120,16 +111,16 @@ GitHub.authorize_url!
 Capture the `code` in your callback route on your server and use it to obtain an access token.
 
 ```elixir
-token = GitHub.get_token!(code: code)
+client = GitHub.get_token!(code: code)
```

Comment:
We have to use `OAuth2` to authenticate with an API. We've created a custom strategy `Foo.Strategy.Bar` by using `OAuth2.Strategy` and defining `get_token!/4` and `get_token/3`.

We pass in a `config` struct to `get_token!/4` which returns the `token`, which we then use to make API calls via `OAuth2.AccessToken.[get|post|...]/4`

```elixir
defmodule Foo.WidgetFetcher do
  def run(config, id) do                                                                                     
    with  {:ok, token}      <- Foo.Strategy.Bar.get_token!(config),                                 
          {:ok, response}   <- OAuth2.AccessToken.get(token,                                                 
                                                      "/widgets/#{id}",               
                                                      [],                                                    
                                                      params: %{extra: true, stuff: true})
    do                                                                                                       
      handle_response({response.status_code, response})                                                      
    end                                                                                                      
    |> case do                                                                                               
      {:ok, results}  -> {:ok, results}                                                                      
      {:not_found}    -> {:not_found, "Unable to find widget"}                                              
      _error          -> {:error, "Unable to retrieve widget from Bar"}                                  
    end                                                                                                      
  end
end

defmodule Foo.Strategy.Bar do
  @moduledoc """
  OAuth2 strategy for Bar
  """
  use OAuth2.Strategy

  def new_client(config) do
    protocol =  unless String.starts_with?(config.host, "http") do
                  "https://"
                end

    OAuth2.Client.new([
      strategy: __MODULE__,
      client_id: config.username,
      client_secret: config.password,
      site: "#{protocol}#{config.host}#{config.root_path}",
      token_url: "#{protocol}#{config.host}/#{auth_root(config.root_path)}/token"
    ])
  end

  def get_token!(config, params \\ [], headers \\ [], opts \\ []) do
    client = new_client(config)
    {token_url, opts} = Keyword.pop(opts, :token_url, client.token_url)
    {site, _opts} = Keyword.pop(opts, :site, client.site)
    client = %{client | site: site, token_url: token_url}

    client
    |> put_param(:auth_scheme, "request_body")
    |> OAuth2.Client.get_token(params, headers)
  end

  def get_token(client, params, headers) do
    client
    |> put_header("Content-type", "application/x-www-form-urlencoded")
    |> OAuth2.Strategy.ClientCredentials.get_token(params, headers)
  end

  def auth_root("/v1"),           do: "oauth"
  def auth_root("/preview1"),     do: "oauthpreview"
  def auth_root("/openpreview1"), do: "oauthopenpreview"
end
                                                                                                     
```

So we actually never use a `client` via `Bar.new`. We load our `config` and then `Foo.WidgetFetcher.run(config, widget_id)`.